dace/frontend/python/replacements.py

```diff
@@ -1582,3 +1592,1287 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
+                visitor, ast_node,
+                "Output parameter for reduction operation {f} has too many "
+                "dimensions (output shape {o}, expected shape {e})."
+                .format(f=ufunc_name, o=out_shape, e=expected_out_shape))
+        if (list(out_shape) != list(expected_out_shape)):
+            raise mem_parser.DaceSyntaxError(
+                visitor, ast_node,
+                "Output parameter for reduction operation {f} has non-reduction"
+                " dimension not equal to the input one (output shape {o}, "
+                "expected shape {e}).".format(
+                    f=ufunc_name, o=out_shape, e=expected_out_shape))
+    else:
+        out_shape = expected_out_shape
+    
+    # No casting needed
+    arg = inputs[0]
+    if isinstance(arg, str):
+        datadesc = sdfg.arrays[arg]
+        result_type = datadesc.dtype
+    elif isinstance(arg, Number):
+        result_type = dtypes.DTYPE_TO_TYPECLASS[type(arg)]
+    elif isinstance(arg, sp.Basic):
+        result_type = _sym_type(arg)
+
+    # Create output data (if needed)
+    outputs = _create_output(sdfg, inputs, outputs, out_shape, result_type,
+                             force_scalar=True)
+    if keepdims:
+        if (len(intermediate_shape) == 1 and intermediate_shape[0] == 1):
+            intermediate_name = sdfg.temp_data_name()
+            sdfg.add_scalar(intermediate_name, result_type, transient=True)
+        else:
+            intermediate_name, _ = sdfg.add_temp_transient(
+                intermediate_shape, result_type)
+    else:
+        intermediate_name = outputs[0]
+    
+    # Validate 'initial' keyword
+    # This is set to be ufunc.identity, when it exists
+    initial = ufunc_impl['initial']
+    if 'initial' in kwargs.keys():
+        # NumPy documentation says that when 'initial' is set to None,
+        # then the first element of the reduction is used. However, it seems
+        # that when 'initial' is None and the ufunc has 'identity', then
+        # ufunc.identity is the default.
+        initial = kwargs['initial'] or initial
+        if initial is None:
+            if isinstance(inputs[0], str) and inputs[0] in sdfg.arrays.keys():
+                inpdata = sdfg.arrays[inputs[0]]
+                # In the input data has more than 1 dimensions and 'initial'
+                # is None, then NumPy uses a different 'initial' value for every
+                # non-reduced dimension.
+                if isinstance(inpdata, data.Array):
+                    state.add_mapped_tasklet(
+                        name=state.label + "_reduce_initial",
+                        map_ranges={
+                            "__i{i}".format(i=i): "0:{s}".format(s=s)
+                            for i, s in enumerate(inpdata.shape)
+                            if i not in axis
+                        },
+                        inputs={
+                            "__inp": dace.Memlet.simple(inputs[0], ','.join([
```

Comment:
could you please use the new dace.Memlet() constructor? it should be easier to work with.